packages/expo-document-picker/ios/DocumentPickerModule.swift

```diff
@@ -0,0 +1,203 @@
+      // TODO: do we need this?
+      //  if (promise.resolver != nil) {
+      //    promise.reject(DocumentPickingInProgressException())
+      //  }
+
+      guard let utilities = self.appContext?.utilities else {
+        throw UtilitiesInterfaceNotFoundException()
+      }
+
+      self.launchDocumentPicker(options: options, promise: promise)
+
+      // TODO: ADD try catch above for catching errors about icloud entitlement
+
+    }.runOnQueue(.main)
+  }
+
+  private var currentPickingContext: PickingContext?
+
+  private func launchDocumentPicker(options: DocumentPickerOptions, promise: Promise) {
+    guard let currentViewController = self.appContext?.utilities?.currentViewController() else {
+      return promise.reject(MissingCurrentViewControllerException())
+    }
+
+    let documentPickerDelegate = DocumentPickerHandler(documentPickingResultHandler: self)
+    let picker: UIDocumentPickerViewController
+
+    let mimeTypes = options.type
+
+    if mimeTypes.isEmpty {
+      return promise.reject(IncorrectTypeArgumentException())
+    }
+
+    let shouldCopyToCacheDirectory = options.copyToCacheDirectory
+
+    if #available(iOS 14, *) {
+      var utTypes: [UTType] = []
+
+      for mimeType in mimeTypes {
+        if let convertedUTType = mimeTypeToUTType(mimeType: mimeType) {
+          utTypes.append(convertedUTType)
+        }
+      }
+      picker = UIDocumentPickerViewController.init(forOpeningContentTypes: utTypes)
+    } else {
+      var utiTypes: [String] = []
+
+      for mimeType in mimeTypes {
+        let convertedUtiType = mimeTypeToUTI(mimeType: mimeType)
+        utiTypes.append(convertedUtiType)
+      }
+
+      picker = UIDocumentPickerViewController.init(documentTypes: utiTypes, in: UIDocumentPickerMode.import)
+    }
+
+    // Because of the way iPad works with Actionsheets such as this one, we need to provide a source view and set it's position.
+    if UIDevice.current.userInterfaceIdiom == UIUserInterfaceIdiom.pad {
+      let currentVCViewFrame = currentViewController.view.frame
+      picker.popoverPresentationController?.sourceRect = CGRect(x: currentVCViewFrame.midX, y: currentVCViewFrame.maxY, width: 0, height: 0)
+    }
+
+    let pickingContext = PickingContext(promise: promise, options: options, documentPickerHandler: documentPickerDelegate)
+
+    picker.delegate = pickingContext.documentPickerHandler
+
+    // Store picking context as we're navigating to the different view controller (starting asynchronous flow)
+    self.currentPickingContext = pickingContext
+    currentViewController.present(picker, animated: true)
+  }
+
+  // MARK: - DocumentPickingResultHandler
+  func didCancelPicking() {
+    self.currentPickingContext?.promise.resolve(["type": "cancel"])
+  }
+
+  func didPickDocument(documentUrl: URL) {
+    guard let options = self.currentPickingContext?.options,
+    let promise = self.currentPickingContext?.promise else {
+      NSLog("Picking operation context has been lost.")
```

Comment:
You can use `log.error` this way:
https://github.com/expo/expo/blob/a0c5fbc58d259de9ccf742a217a0be266aeaa3ec/packages/expo-image-picker/ios/ImagePickerModule.swift#L178